perftest/PerfTest.cpp

```diff
@@ -0,0 +1,275 @@
+            int sleepms = std::min(std::max(1,elapsedms), 1000);
+            std::this_thread::sleep_for(std::chrono::milliseconds(sleepms));
+        }
+    } while (!found && !timedout);
+
+    double seconds = std::chrono::duration_cast<std::chrono::milliseconds>(now - _waitingStart).count()/1000.0;
+    if (found) {
+        LOG_DBG("PerfTest waitForMessage Found message: " + substring + " after " + std::to_string(seconds) + "s");
+    } else {
+        abort("Timed out waiting for message \"" + substring + "\" after " + std::to_string(seconds) + "s");
+    }
+}
+
+void PerfTest::waitForIdle(size_t timeout)
+{
+    LOG_TRC("PerfTest waitForIdle");
+    sendMessage("uno .uno:WaitForIdle");
+    waitForMessage("WaitForIdle",timeout);
+}
+
+void PerfTest::sendMessage(const std::string &message)
+{
+    LOG_TRC("PerfTest sendMessage: " + message);
+    _handler->sendMessage(message);
+}
+
+void PerfTest::sleep(unsigned int millis)
+{
+    LOG_DBG("PerfTest sleep: " + std::to_string(millis) + "ms");
+    std::this_thread::sleep_for(std::chrono::milliseconds(millis));
+}
+
+void PerfTest::disconnect()
+{
+    LOG_DBG("PerfTest disconnect");
+    _handler->shutdown();
+}
+
+CyclePerfTest::CyclePerfTest(const std::string &name, const std::string &server) :
+    PerfTest(name, server)
+{
+}
+
+void CyclePerfTest::startMeasurement()
+{
+    LOG_DBG("CyclePerfTest startMeasurement");
+    PerfTest::startMeasurement();
+    pid_t pid;
+    pid = vfork();
+    if (pid < 0) {
+        perror("fork");
+        Util::forcedExit(EX_SOFTWARE);
+    } else if (pid == 0) {
+        LOG_DBG("Starting perf");
+        std::string pid_str = "--pid="+std::to_string(getCoolwsdPid());
+        std::string output_str = "--output="+_fileName+".data";
+        LOG_DBG("Starting perf"<<"perf"<<"record"<<"-s"<<"-e"<<"cycles"<<"--freq=1000"<<"--call-graph"<<"dwarf"<<pid_str.c_str()<<output_str.c_str());
+        execlp("perf","perf","record","-s","-e","cycles","--freq=1000","--call-graph","dwarf",pid_str.c_str(),output_str.c_str(),(char *)NULL);
+    } else {
+        child_pid = pid;
+    }
+}
+
+void CyclePerfTest::stopMeasurement()
+{
+    LOG_DBG("CyclePerfTest stopMeasurement");
+
+    // Stop measurement
+    PerfTest::stopMeasurement();
+    kill(child_pid,SIGINT);
+    waitpid(child_pid, nullptr, 0);
+    addResult("dataFile",_fileName+".data");
+
+    // Get result
+    // Use perf report to get summary statistics from the data file
+    // Use grep and awk to get the cpu cycle counts from the report output
+    // Use paste to format list of numbers for bc
+    // Use bc to sum numbers. bc, as an arbitrary precision calculator, is
+    // necessary because cycle counts can be >2^31 (too big for awk's sum)
+    std::string command = "perf report -i "+_fileName+".data --no-inline | grep \"# Event count (approx.): \" | awk '{print $5}' | paste -s -d+ | bc";
```

Comment:
Ah, this is not system() but needs similar fixing to avoid start programming in bash-in-C++.

Reply:
One more tip here: you probably only need to execute 'perf report' here. If you use code from common/Util.hpp and StringVector::tokenize(), then you can avoid the external grep / awk / etc calls.